libbeat/processors/cache/cache.go

```diff
@@ -302,7 +313,9 @@ func (p *cache) deleteFor(event *beat.Event) (err error) {
 }
 
 func (p *cache) Close() error {
-	p.cancel()
+	if p.cancel != nil {
```

Comment:
I think we might have a race condition here.  Is it possible that `Close` could be called before `SetPaths`? If so we I think we'll need a `sync.Mutex` or `sync.Once` to make sure we either go in the right order or don't do `SetPaths` if `Close` is called first.

Reply:
Hmm, perhaps a filebeat input can call it while filebeat is exiting. Given that this is problem is going to be shared across all processors, it could perhaps be better to implement it in safe_processor.go